### PR TITLE
[16.0][IMP] queue_job: introduce a new hook mechanism to handle queues in max retries reached case

### PR DESCRIPTION
**Pull request by QuocDuong1306**
When a job has been tried for max_retries but still fails, final FailedJobError is raised and job is set to Failed.
This PR enables developers to run a specific hook method when this happens.

### CODE REVIEW

File: queue_job/job.py [hunk 1]
```diff
@@ -527,6 +527,14 @@ def perform(self):
...
                 raise
             elif self.retry >= self.max_retries:
+                hook = f"{self.method_name}_on_max_retries_reached"
+                if hasattr(self.recordset, hook):
+                    recordset = self.recordset.with_context(
+                        job_uuid=self.uuid, exc_info=self.exc_info
+                    )
+                    getattr(recordset, hook)(*tuple(self.args))
+                    # Should be set_done after perform
+                    return
```
Comment (nilshamerlinck) [inline queue_job/job.py]: I'm confused about that part, why return?
we still want to raise final FailedJobError
to set the job as failed
see [here](https://github.com/OCA/queue/blob/373cbca82b9dd942184e3926e1c67ca25b6aa373/queue_job/controllers/main.py#L135)

File: queue_job/job.py [hunk 2]
```diff
@@ -527,6 +527,14 @@ def perform(self):
             elif not self.max_retries:  # infinite retries
                 raise
             elif self.retry >= self.max_retries:
+                hook = f"{self.method_name}_on_max_retries_reached"
+                if hasattr(self.recordset, hook):
+                    recordset = self.recordset.with_context(
+                        job_uuid=self.uuid, exc_info=self.exc_info
+                    )
+                    getattr(recordset, hook)(*tuple(self.args))
```
Comment (nilshamerlinck) [inline queue_job/job.py]: what are pro/cons of passing self.args vs passing job?

**Comment by nilshamerlinck** [comment]
thanks for the first iteration!

> Currently, when the queue reaches the max_retries, the FailedJobError will be raised.
> In some cases, some specific models don't want to get that Error and want to apply new behaviors, this commit will provide the hook to support those cases.
> After using the hook, the queue job won't annoy anymore and stay as final state (Done).

I would suggest a smaller description instead, no need to get into too much details:

When a _job_ has been tried for `max_retries` but still fails, final FailedJobError is raised and job is set to Failed.
This PR enables developers to run a specific hook method when this happens.

File: queue_job/job.py (hunk 1 above)
Comment (QuocDuong1306) [reply queue_job/job.py]: Ah my bad, I misunderstood, I thought that the failed job would be retrieved and re-start

File: queue_job/job.py (hunk 2 above)
Comment (QuocDuong1306) [reply queue_job/job.py]: We need to get exc records to run the action to put it as final state

If passing job, not have any relation between the job and related records, maybe we need to search with  `func_string` to get the exc records --> performance

btw, get more similar with the origin methods:
ex: `def exchange_receive(self, exchange_record)` and `def exchange_receive_on_max_retries_reached(self, exchange_records)`

Comment (nilshamerlinck) [reply queue_job/job.py]: > if passing job, not have any relation between the job and related records

but with `getattr(recordset, hook)` you already get a method which is bound to the related records

Comment (QuocDuong1306) [reply queue_job/job.py]: In the origin context, I assumed that Jobs have been applied on `backend`

For now, I'm clear, Thank nils

File: queue_job/job.py [hunk 3]
```diff
@@ -527,6 +527,12 @@ def perform(self):
             elif not self.max_retries:  # infinite retries
                 raise
             elif self.retry >= self.max_retries:
+                hook = f"{self.method_name}_on_max_retries_reached"
+                if hasattr(self.recordset, hook):
+                    recordset = self.recordset.with_context(
+                        job_uuid=self.uuid, exc_info=self.exc_info
+                    )
+                    getattr(recordset, hook)()
```
Comment (nilshamerlinck) [inline queue_job/job.py]:
```suggestion
                    try:
                        getattr(recordset, hook)()
                    except Exception as ex:
                        _logger.debug(
                            "exception on %s:%s() for job uuid: %s",
                            self.recordset,
                            hook,
                            self.uuid,
                            ex
                        )                
```

Comment (nilshamerlinck) [reply queue_job/job.py]: last minute thought: better be safe

**Review by nilshamerlinck — APPROVED** [review]
pre-approving

can you please also change PR/commit name into shorter:

"[IMP] queue_job: run specific hook method after max_retries"

**Pull request closed without merging**